File: go/nodule.go
package specimen

import (
	"errors"
	"fmt"

	"github.com/ditrit/specimen/go/focustree"
	"github.com/ditrit/specimen/go/syaml"
	"gopkg.in/yaml.v3"
)

func (n *Nodule) GetLocation() string {
	return fmt.Sprintf("%s:%d:%d", n.FilePath, n.YamlNode.Line, n.YamlNode.Column)
}

func (n Nodule) Errorf(format string, a ...interface{}) error {
	info := fmt.Sprintf(format, a...)
	return fmt.Errorf("(%s): %s", n.GetLocation(), info)
}

func NewNoduleFromFile(file File) (n Nodule, err error) {
	n.FilePath = file.Path

	document := &yaml.Node{}

	err = yaml.Unmarshal(file.Content, document)
	if err != nil {
		return
	}

	// mapping
	n.YamlNode, err = syaml.EnterDocument(document)
	if err != nil {
		return
	}
	if !syc.IsMapping(n.YamlNode) {
		err = fmt.Errorf("the root of the document must be a yaml mapping")
		return
	}

	n.InitializeTree()

	return
}

// The initialization creates all the nodules corresponding to the mapping nodes of the yaml tree, except for the PENDING nodes. It fills the fields Flag, HasContentKey and Children. **It expects YamlNode and FilePath to be already set**, and it sets YamlNode and FilePath for its children.
func (n *Nodule) InitializeTree() (err error) {
	if !syc.IsMapping(n.YamlNode) {
		err = fmt.Errorf("the root of the document must be a yaml mapping")
		return
	}
	flagNode := syc.MapTryGetValue(n.YamlNode, "flag")
	if flagNode != nil {
		n.Flag = readFlag(flagNode)
	}

	if n.Flag == focustree.Skip {
		return
	}

	contentNode := syc.MapTryGetValue(n.YamlNode, "content")
	if contentNode != nil {
		n.HasContentKey = true
		for _, childYamlNode := range contentNode.Content {
			child := Nodule{
				FilePath: n.FilePath,
				YamlNode: childYamlNode,
			}
			child.InitializeTree()
			n.Children = append(n.Children, child)
		}
	}

	return
}

// Populate fills the DataMatrix and DataOrder fields from the Yaml data
func (n *Nodule) Populate(dataMatrix map[string][]string, dataOrder []string) (err error) {
	// Todo: improve the performance by detecting all the cases where creating a copy of the dataMatrix and dataOrder is unneccessary
	if len(n.YamlNode.Content) == 0 {
		return nil
	}

	// Cloning dataMatrix and dataOrder
	n.DataMatrix = map[string][]string{}
	for k, v := range dataMatrix {
		n.DataMatrix[k] = v
	}
	n.DataOrder = make(
		[]string,
		len(dataOrder),
		len(dataOrder)+len(n.YamlNode.Content)/2,
	)
	copy(n.DataOrder, dataOrder)

	errorSlice := []error{}

	for k := 0; k < len(n.YamlNode.Content)/2; k += 1 {
		key := n.YamlNode.Content[2*k]
		if key.Value == "content" {
			continue
		}
		value := n.YamlNode.Content[2*k+1]
		if value.Kind == yaml.SequenceNode {
			slice := []string{}
			for _, entry := range value.Content {
				slice = append(slice, entry.Value)
			}
			n.DataMatrix[key.Value] = slice
			n.DataOrder = append(n.DataOrder, key.Value)
		} else if value.Kind == yaml.ScalarNode {
			n.DataMatrix[key.Value] = []string{value.Value}
			n.DataOrder = append(n.DataOrder, key.Value)
			// /!\ TODO: extract DataMatrix&DataOrder to their own module and
			// perform a check to see if the entry is already in the map, so
			// as to avoid duplicate entries in the DataOrder slice.
		} else {
			errorSlice = append(errorSlice, n.Errorf("Unexpected node kind [%d] for key [%s]", value.Kind, key.Value))
		}
	}

	// Recursing over children
	for k := range n.Children {
		errorSlice = append(errorSlice, n.Children[k].Populate(n.DataMatrix, n.DataOrder))
	}

	err = errors.Join(errorSlice...)

	return
}

// TODO: make it so inserting a key in n.DataOrder checks and when present, removes the older instance of the key.

func (n *Nodule) NewResolveDataMatrixIterator() func() Dict {
	// reverse the dataOrder so that we iterate quickly on the latest keys, and
	// more slowly in the earlier keys
	dataOrder := make([]string, len(n.DataOrder))
	for k, v := range n.DataOrder {
		dataOrder[len(n.DataOrder)-1-k] = v
	}

	// Calculate the total number of combinations and the intermediate slice
	// sizes
	totalCombinations := 1
	sizeSlice := make([]int, 0, len(n.DataOrder))
	for _, key := range dataOrder {
		set := n.DataMatrix[key]
		totalCombinations *= len(set)
		sizeSlice = append(sizeSlice, totalCombinations)
	}

	// The indexSlice traks the progress of values through every set
	indexSlice := make([]int, len(dataOrder))

	// The combination is the variable that will be updated and returned by the
	// iterator
	combination := Dict{}
	// Initialize the combination
	for k, key := range dataOrder {
		combination[key] = n.DataMatrix[key][indexSlice[k]]
		// ^ Note that in this loop, indexSlice[k] is actually always 0
	}

	// Create a closure-based iterator function
	index := 0
	return func() Dict {
		if index == 0 {
			index += 1
			return combination
		} else if index == totalCombinations {
			index += 1
			return nil
		}

		// Go through the keys to find which one is affected by the index change
		for k, key := range dataOrder {
			if index%sizeSlice[k] > 0 {
				// bump the identified index
				indexSlice[k] += 1
				// update the combination entry corresponding to the identified key
				combination[key] = n.DataMatrix[key][indexSlice[k]]
				break
			}
		}
		index += 1
		return combination
	}
}

File: rust/src/nodule.rs
use crate::file;
use crate::flag;
use multistringmap::MultiStringMap;
use std::rc::Rc;
use yaml;

#[derive(Clone, Debug)]
pub struct Nodule<'a> {
    pub node: &'a yaml::Yaml,
    pub flag: focustree::Flag,
    pub is_leaf: bool,
    pub file_path: Rc<str>,
    pub data_matrix: MultiStringMap,
    pub children: Box<[Nodule<'a>]>,
}

impl<'a> Nodule<'a> {
    // Associated functions
    pub fn parse_file(file: &file::File, store: &'a mut Box<Vec<yaml::Yaml>>) -> Vec<Nodule<'a>> {
        let file_path = Rc::from(file.path.to_owned());

        let mut document_vec = match yaml::YamlLoader::load_from_str(&file.content) {
            Ok(v) => v,
            Err(e) => {
                println!("{}: {}", file.path, e);
                vec![]
            }
        };

        std::mem::swap(&mut document_vec, store);

        store
            .iter()
            .map(|node| {
                match node.data {
                    yaml::YamlData::Mapping(_) => {}
                    _ => panic!("The root node of the YAML test data file must be a mapping."),
                }

                let mut data_matrix = MultiStringMap::new();
                data_matrix.0.insert(
                    Box::from("file_path"),
                    Rc::new([Box::from(file.path.clone())]),
                );

                let mut n = Nodule {
                    node: &node,
                    flag: focustree::Flag::None,
                    is_leaf: true,
                    file_path: Rc::clone(&file_path),
                    data_matrix,
                    children: Box::new([]),
                };

                n.initalize_tree();

                n
            })
            .collect()
    }

    // Methods
    pub fn get_location(&self) -> Box<str> {
        format!(
            "{}:{}:{}",
            self.file_path, self.node.position.line, self.node.position.column
        )
        .into()
    }

    // The initialization creates all the nodules which correspond to the mapping nodes of the yaml tree, except for the PENDING nodes. It fills the fields `flag`, `has_content_key` and `children`. **It expects YamlNode and FilePath to be already set**, and it sets YamlNode and FilePath for its children.
    fn initalize_tree(&mut self) {
        match self.node.data {
            yaml::YamlData::Mapping(_) => {}
            _ => panic!("the content descendant nodes must be yaml mappings"),
        }

        let flag_node = &self.node.data["flag"];
        if *flag_node != yaml::BAD_VALUE {
            self.flag = flag::read_flag(&flag_node);
        }
        if self.flag == focustree::Flag::Skip {
            return;
        }

        let content_node = &self.node.data["content"];
        if *content_node != yaml::BAD_VALUE {
            self.is_leaf = false;
            if let yaml::YamlData::List(ref yaml_vec) = content_node.data {
                self.children = yaml_vec
                    .iter()
                    .map(|node| {
                        let mut n = Nodule {
                            node,
                            flag: focustree::Flag::None,
                            is_leaf: true,
                            file_path: Rc::clone(&self.file_path),
                            children: Box::new([]),
                            data_matrix: MultiStringMap::new(),
                        };
                        n.initalize_tree();
                        n
                    })
                    .collect();
            } else {
                panic!(
                    "the value associated with the content keyword must be a sequence of mappings"
                )
            }
        }
    }

    pub fn populate(
        &mut self,
        data_matrix: &MultiStringMap,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let data = match self.node.data {
            yaml::YamlData::Mapping(ref m) => m,
            _ => self.panic("the content descendant nodes must be yaml mappings"),
        };

        self.data_matrix = data_matrix.clone();

        for (key, value) in data.iter() {
            let key = match key.data {
                yaml::YamlData::String(ref s) => s,
                _ => self.panic("the keys of the mapping nodes must be strings"),
            };
            if key == "flag" || key == "content" || key == "about" {
                continue;
            }

            let value_vector: Vec<Box<str>> = match value.data {
                yaml::YamlData::String(ref s) => vec![Box::from(s.to_owned())],
                yaml::YamlData::List(ref a) => {
                    if a.len() == 0 {
                        self.panic("when the values of the mapping nodes is a sequence, it must not be empty.")
                    }
                    a.iter().map(|v| match v.data {
                        yaml::YamlData::String(ref s) => Box::from(s.to_owned()),
                        _ => self.panic("when the values of the mapping nodes is a sequence, it must be a sequence of strings only."),
                    }).collect()
                }
                _ => self.panic(&format!(
                    "the values of mapping nodes must be strings or sequences. (key: {:?})",
                    key
                )),
            };

            self.data_matrix
                .0
                .insert(key.to_owned().into_boxed_str(), Rc::from(value_vector));
        }

        for child in self.children.iter_mut() {
            child.populate(&self.data_matrix)?;
        }

        Ok(())
    }

    fn panic(&self, message: &str) -> ! {
        panic!(
            "{} ({}:{} with data: {:?})",
            message, self.file_path, self.node.position, self.node.data
        )
    }
}
